### DISCUSSION TITLE
How best to accelerate sparse scene queries in raytracing?

### DISCUSSION THREAD

**Discussion by rigGitLikeYouDigGit** [taichi-dev/taichi#6093] [Q&A] [answered]
Hi, I'm struggling to understand how to use Taichi's sparse systems for a use like raytracing. I'm also new to Taichi in general, so apologies if I misunderstand anything. 

I'm using this project as reference https://github.com/lyd405121/ti-raytrace - while it seems excellent for everything else, for space acceleration, it just builds and balances a BVH tree in https://github.com/lyd405121/ti-raytrace/blob/main/accel/LBvh.py

This is obviously fine, but I'd love to leverage the builtin SNode system, since it's one of Taichi's main advantages. 

In this case, my plan is:
-create some variation of SNode hierarchy (probably start out emulating VDB branching factor) - with leaf cells holding integers
-use those integers as pointers back into a store of data structs, with each struct holding a list of all the scene primitives intersecting that leaf node (of which we might have any number).

Then during ray cast, run normal bbox-ray intersection checks only against active nodes, recursing down to leaves, and finally checking each primitive contained in a leaf individually.

Is having any number of primitives per leaf cell legal? Would it be better to define a regular array length, with a maximum number of primitives that can be held?

I'd appreciate any advice you can give on this, thanks.

**Comment by bobcao3** [comment, marked as answer]
Taichi's SNode is more or less designed for spatially sparse objects on a regular grid. Triangle or mesh primitive ray-tracing acceleration does not map well to a regular grid (think nyqusit frequency as a fundamental limitation), thus you don't see people use regular octree or VDB similar systems for general scene data. It will be pretty good for volume data which is usually sampled on a regular grid.

**Reply by rigGitLikeYouDigGit** [reply]
I see, thanks - I'm not sure I understand the distinction between raytracing and more grid-suitable applications though. Isn't the act of looking up the nearest primitives to a ray, analoguous to looking up nearest neighbours in something like a particle sim? 
Also can you elaborate more on the Nyquist frequency? I've only ever heard that in relation to continuous signals, how does it affect space acceleration like this?

**Reply by bobcao3** [reply]
because geometry is unlike volumes or particles in they way they are distributed in space. Particles and volumes are more or less uniformly distributed and they have a low or at least consistent signal frequency (i.e. you can put them onto a grid and there won't be a lot of paricles in one cell and very few of them in other). Triangles and mesh in nature are not like this. They vary in size and shape, they span multiple cells usually, and their occurrence are not regular, thus very hard to put onto a regular grid
